### PR TITLE
Adds two modes for BuildMode

### PR DESCRIPTION
**Pull request by AnywayFarus**
## About The Pull Request

Adds two modes for BuildMode, with personal icons

- Tweak Components Rating - for easier upgrade of machinery.
- ProcCall - for fast ProcCall on objects.

## Why It's Good For The Game

This can help the staff in carrying out all sorts of events, all sorts of in-game tests, because it provides a convenient and nimble toolkit.

## Changelog

:cl: Vishenka0704
add: Added "ProcCall" & "Tweak Components Rating" modes in buildmode
/:cl:

### CODE REVIEW

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+		return
+	else
+		if(object.component_parts)
+			for(var/obj/item/stock_parts/P in object.component_parts)
+				P.rating = rating
+			object.RefreshParts()
+			
+			SSblackbox.record_feedback("nested tally", "admin_toggle", 1, list("Machine Upgrade", "[rating]")) //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
+		else
+			to_chat(target_client, span_warning("That machinery don't have components"))
```
Comment (san7890) [inline code/modules/buildmode/submodes/tweakcomps.dm:32]:
```suggestion
			to_chat(target_client, span_warning("This machinery doesn't have components!"))
```

Grammar.

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/datum/buildmode_mode/tweakcomps
+	key = "tweakcomps"
+	var/rating = null
+
+/datum/buildmode_mode/tweakcomps/show_help(client/target_client)
+	to_chat(target_client, span_notice("***********************************************************\n\
+		Right Mouse Button on buildmode button = Choose components rating\n\
+		Left Mouse Button on machinery = Sets components choosen rating.\n\
```
Comment (san7890) [inline code/modules/buildmode/submodes/tweakcomps.dm:8]:
```suggestion
		Right Mouse Button on buildmode button = Choose the rating of the components.\n\
		Left Mouse Button on machinery = Sets the chosen rating of the components on the machinery.\n\
```

Grammar, this one might be a bit lengthy though?

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	rating_to_choose = text2num(rating_to_choose)
+	if(!isnum(rating_to_choose))
+		tgui_alert(target_client, "Input a number.")
+		return
+	else
+		rating = rating_to_choose
+
+/datum/buildmode_mode/tweakcomps/handle_click(client/target_client, params, obj/machinery/object)
+	if(!ismachinery(object))
+		to_chat(target_client, span_warning("This object is not machinery!"))
```
Comment (san7890) [inline code/modules/buildmode/submodes/tweakcomps.dm:22]:
```suggestion
		to_chat(target_client, span_warning("This isn't machinery!"))
```

We can afford to be a little less verbose here, I feel.

File: code/modules/buildmode/submodes/proccall.dm [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+	if(!proc_args)
+		return
+
+/datum/buildmode_mode/proccall/handle_click(client/target_client, params, datum/object as null|area|mob|obj|turf)
+	if(!proc_name || !proc_args)
+		tgui_alert(target_client, "Undefined ProcCall or arguments.")
+		return
+	
+	if(!hascall(object, proc_name))
+		to_chat(target_client, "<font color='red'>Error: callproc_datum(): type [object.type] has no proc named [proc_name].</font>", confidential = TRUE)
```
Comment (san7890) [inline code/modules/buildmode/submodes/proccall.dm:30]:
```suggestion
		to_chat(target_client, "span_warning(Error: callproc_datum(): type [object.type] has no proc named [proc_name].)", confidential = TRUE)
```

No need to manually set the color when we have spans that do the heavy lifting for us.

**Review by san7890 — COMMENTED** [review]
Minor things.

**Comment by Arturlang** [comment]
Any chance of adding zone saving from bee or somesuch?

**Comment by AnywayFarus** [comment]
> Any chance of adding zone saving from bee or somesuch?

I have no idea what you mean to be honest :)

**Comment by Arturlang** [comment]
@AnywayFarus If you're not being sarcastic, I mean using build mode for saving certain parts of the map to client's disk as a DMM. I'm fairly sure bee already has this
https://github.com/BeeStation/BeeStation-Hornet/pull/1522

**Comment by AnywayFarus** [comment]
> @AnywayFarus If you're not being sarcastic, I mean using build mode for saving certain parts of the map to client's disk as a DMM. I'm fairly sure bee already has this [BeeStation/BeeStation-Hornet#1522](https://github.com/BeeStation/BeeStation-Hornet/pull/1522)

maybe in another PR :(

File: code/modules/buildmode/submodes/proccall.dm [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+/datum/buildmode_mode/proccall/handle_click(client/target_client, params, datum/object as null|area|mob|obj|turf)
+	if(!proc_name || !proc_args)
+		tgui_alert(target_client, "Undefined ProcCall or arguments.")
+		return
+	
+	if(!hascall(object, proc_name))
+		to_chat(target_client, "span_warning(Error: callproc_datum(): type [object.type] has no proc named [proc_name].)", confidential = TRUE)
+		return
+
+	if(!object || !is_valid_src(object))
```
Comment (ShizCalev) [inline code/modules/buildmode/submodes/proccall.dm:33]: is_valid_src handles !object

File: code/modules/buildmode/submodes/proccall.dm [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+	if(!proc_args)
+		return
+
+/datum/buildmode_mode/proccall/handle_click(client/target_client, params, datum/object as null|area|mob|obj|turf)
+	if(!proc_name || !proc_args)
+		tgui_alert(target_client, "Undefined ProcCall or arguments.")
+		return
+	
+	if(!hascall(object, proc_name))
+		to_chat(target_client, "span_warning(Error: callproc_datum(): type [object.type] has no proc named [proc_name].)", confidential = TRUE)
```
Comment (SuperNovaa41) [inline code/modules/buildmode/submodes/proccall.dm:30]:
```suggestion
		to_chat(target_client, span_warning("Error: callproc_datum(): type [object.type] has no proc named [proc_name].)", confidential = TRUE)
```

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]

File: code/modules/buildmode/submodes/proccall.dm (hunk 6 above)
Comment (AnywayFarus) [reply code/modules/buildmode/submodes/proccall.dm:30]: hah.

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+		Left Mouse Button on machinery = Sets the chosen rating of the components on the machinery.\n\
+		***********************************************************"))
+
+/datum/buildmode_mode/tweakcomps/change_settings(client/target_client)
+	var/rating_to_choose = input(target_client, "Enter number of rating", "Number", "1") 
+	rating_to_choose = text2num(rating_to_choose)
+	if(!isnum(rating_to_choose))
+		tgui_alert(target_client, "Input a number.")
+		return
+	else
```
Comment (tralezab) [inline code/modules/buildmode/submodes/tweakcomps.dm:17]: if the function ends, you don't need an else case

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 8]
```diff
@@ -0,0 +1,33 @@
+/datum/buildmode_mode/tweakcomps
+	key = "tweakcomps"
+	var/rating = null
```
Comment (tralezab) [inline code/modules/buildmode/submodes/tweakcomps.dm:3]: document this var

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+		tgui_alert(target_client, "Input a number.")
+		return
+	else
+		rating = rating_to_choose
+
+/datum/buildmode_mode/tweakcomps/handle_click(client/target_client, params, obj/machinery/object)
+	if(!ismachinery(object))
+		to_chat(target_client, span_warning("This isn't machinery!"))
+		return
+	else
```
Comment (tralezab) [inline code/modules/buildmode/submodes/tweakcomps.dm:24]: same as above, return ends a function so don't put an else case

File: code/modules/buildmode/submodes/tweakcomps.dm [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+		if(object.component_parts)
+			for(var/obj/item/stock_parts/P in object.component_parts)
+				P.rating = rating
+			object.RefreshParts()
+			
+			SSblackbox.record_feedback("nested tally", "admin_toggle", 1, list("Machine Upgrade", "[rating]")) //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
+		else
+			to_chat(target_client, span_warning("This machinery doesn't have components!"))
+			return
```
Comment (tralezab) [inline code/modules/buildmode/submodes/tweakcomps.dm:33]:
```suggestion
		if(!object.component_parts)
			to_chat(target_client, span_warning("This machinery doesn't have components!"))
			return
		for(var/obj/item/stock_parts/P in object.component_parts)
			P.rating = rating
		object.RefreshParts()
		
		SSblackbox.record_feedback("nested tally", "admin_toggle", 1, list("Machine Upgrade", "[rating]")) //If you are copy-pasting this, ensure the 2nd parameter is unique to the new proc!
```

**Review by tralezab — CHANGES_REQUESTED** [review]
some stuff

**Pull request merged**